ReserveSlot and context cancellation

## Expected Behavior

`SlotSupplier.ReserveSlot` returns `context.Canceled` when its context is cancelled, so that the poller goroutine calling it can exit during worker shutdown.

That is what the interface documents:

```go
// ReserveSlot is called before polling for new tasks. The implementation should block until
// a slot is available, then return a permit to use that slot. ...
//
// Any returned error besides context.Canceled will be logged and the function will be retried.
```

and it is what `FixedSizeSlotSupplier` does (`internal/tuning.go`), via `f.sem.Acquire(ctx, 1)`.

## Actual Behavior

`ResourceBasedSlotSupplier.ReserveSlot` appears not to observe cancellation while the resource controller is declining slots. The goroutine keeps looping, and does not return, even long after the context is done.

Looking at `internal/resource_tuner.go` (`ReserveSlot`), the only `ctx.Done()` check is inside two conditions:

```go
for {
    if info.NumIssuedSlots() < r.options.MinSlots {
        return &SlotPermit{}, nil
    }
    if r.options.RampThrottle > 0 {                                     // (1)
        r.lastIssuedMu.Lock()
        mustWaitFor := r.options.RampThrottle - time.Since(r.lastSlotIssuedAt)
        r.lastIssuedMu.Unlock()
        if mustWaitFor > 0 {                                            // (2)
            select {
            case <-time.After(mustWaitFor):
            case <-ctx.Done():                                          // <- the only ctx check
                return nil, ctx.Err()
            }
        }
    }

    maybePermit := r.TryReserveSlot(info)
    if maybePermit != nil {
        return maybePermit, nil
    }
    time.Sleep(10 * time.Millisecond)                                   // (3) not ctx-aware
}
```

If I am reading this correctly, cancellation is only observed while the ramp throttle happens to be making the caller wait. When it is not, the loop has no cancellation point, and `time.Sleep` at (3) does not provide one either. Two ways to get there:

- **(1) is false**: `RampThrottle == 0`, which is the default for workflow slot suppliers (`DefaultWorkflowResourceBasedSlotSupplierOptions`). The whole block is skipped.
- **(2) is false**: `RampThrottle > 0` but the last slot was issued longer than `RampThrottle` ago, so `mustWaitFor <= 0`. This seems to be the steady state whenever slots are not being issued.

The loop then only exits by obtaining a permit. But `TryReserveSlot` returns `nil` whenever `pidDecision` declines, which is exactly the sustained-resource-pressure case (e.g. memory at or above `MemTargetPercent`) that the tuner exists to handle.

The reason I think this matters rather than being merely untidy: `reserveSlotAsync` (`internal/internal_worker_base.go`) runs `ReserveSlot` in a goroutine registered on the worker's stop WaitGroup:

```go
bw.stopWG.Add(1)
go func() {
    defer bw.stopWG.Done()
    s, err := bw.slotSupplier.ReserveSlot(ctx, &reservationData)
    ...
```

and `baseWorker.Stop()` waits on that WaitGroup via `awaitWaitGroup(&bw.stopWG, bw.options.stopTimeout)`. `runPoller` does cancel the context on its way out (`defer cancelfn()`), so the cancellation signal is delivered. It just does not appear to be acted on.

So under resource pressure, my reading is that the goroutine never finishes, `stopWG` never drains, and:

- `Stop()` reports `"Worker graceful stop timed out."` even though nothing is actually processing a task (with the default `WorkerStopTimeout` of 0 this returns immediately, so the symptom is the log line plus the leak; if `WorkerStopTimeout` is set, `Stop()` blocks for its full duration).
- the goroutine outlives the worker, spinning on the 10ms loop and calling the `SysInfoProvider` about 180 times a second, for the life of the process.

## Steps to Reproduce the Problem

Self-contained test, no server needed. It stands up a `ResourceBasedSlotSupplier` whose `SysInfoProvider` reports memory above the target (so `pidDecision` always declines), calls `ReserveSlot` in a goroutine, cancels the context, and then checks whether the call returns.

<details>
<summary><code>internal/resource_tuner_cancel_test.go</code></summary>

```go
package internal

import (
	"context"
	"sync/atomic"
	"testing"
	"time"

	enumspb "go.temporal.io/api/enums/v1"
	"go.temporal.io/sdk/internal/common/metrics"
	ilog "go.temporal.io/sdk/internal/log"
	"go.temporal.io/sdk/log"
)

type cancelInfo struct{ calls atomic.Int64 }

func (i *cancelInfo) TaskQueue() string                    { return "tq" }
func (i *cancelInfo) TaskQueueKind() enumspb.TaskQueueKind { return enumspb.TASK_QUEUE_KIND_NORMAL }
func (i *cancelInfo) WorkerBuildId() string                { return "" }
func (i *cancelInfo) WorkerIdentity() string               { return "" }
func (i *cancelInfo) NumIssuedSlots() int                  { i.calls.Add(1); return 10 } // >= MinSlots
func (i *cancelInfo) Logger() log.Logger                   { return ilog.NewNopLogger() }
func (i *cancelInfo) MetricsHandler() metrics.Handler      { return metrics.NopHandler }

// Memory above the target, so pidDecision always declines and TryReserveSlot always returns nil.
type underPressure struct{}

func (underPressure) MemoryUsage(*SysInfoContext) (float64, error) { return 0.95, nil }
func (underPressure) CpuUsage(*SysInfoContext) (float64, error)     { return 0.95, nil }

func TestReserveSlotHonoursContextCancellation(t *testing.T) {
	for _, tc := range []struct {
		name string
		ramp time.Duration
	}{
		{"RampThrottle=0 (workflow default)", 0},
		{"RampThrottle=50ms (activity default)", 50 * time.Millisecond},
	} {
		t.Run(tc.name, func(t *testing.T) {
			opts := DefaultResourceControllerOptions()
			opts.MemTargetPercent = 0.8
			opts.CpuTargetPercent = 0.9
			opts.InfoSupplier = underPressure{}

			supplier, err := NewResourceBasedSlotSupplier(
				NewResourceController(opts),
				ResourceBasedSlotSupplierOptions{
					MinSlots:      0,
					MaxSlots:      1000,
					RampThrottle: tc.ramp,
				},
			)
			if err != nil {
				t.Fatal(err)
			}

			info := &cancelInfo{}
			ctx, cancel := context.WithCancel(context.Background())

			returned := make(chan error, 1)
			go func() {
				_, err := supplier.ReserveSlot(ctx, info)
				returned <- err
			}()

			time.Sleep(100 * time.Millisecond)
			cancel() // this is what worker.Stop() does to the poller's context

			select {
			case err := <-returned:
				t.Logf("ReserveSlot returned %v", err)
			case <-time.After(1 * time.Second):
				before := info.calls.Load()
				time.Sleep(500 * time.Millisecond)
				t.Errorf(
					"ReserveSlot did not return 1s after cancel; still spinning (%d more iterations in the last 500ms)",
					info.calls.Load()-before,
				)
			}
		})
	}
}
```

</details>

Result on `main` (`8e2c89c7`), both subtests:

```text
--- FAIL: TestReserveSlotHonoursContextCancellation
    --- FAIL: TestReserveSlotHonoursContextCancellation/RampThrottle=0_(workflow_default)
        ReserveSlot did not return 1s after cancel; still spinning (92 more iterations in the last 500ms)
    --- FAIL: TestReserveSlotHonoursContextCancellation/RampThrottle=50ms_(activity_default)
        ReserveSlot did not return 1s after cancel; still spinning (92 more iterations in the last 500ms)
```

## What I am unsure about, and what I am asking

I found this while working on #2443 / #2444 (a separate problem in the same function), so I have been staring at this code, but I have not run it against a real worker under real memory pressure. Before I put up a PR I would rather check the reading with someone who knows the intent here:

1. **Is `ReserveSlot` meant to honour cancellation in this path at all?** The interface comment and `reserveSlotAsync`'s `errors.Is(err, context.Canceled)` check both suggest yes, and `FixedSizeSlotSupplier` does. But I may be missing a mechanism that stops the goroutine some other way.
2. **Is a leaked poller goroutine here actually harmful in practice**, or is it expected to be reclaimed somehow? My concern is that it survives `Stop()` and keeps polling the `SysInfoProvider` (gopsutil syscalls, in the usual configuration) for the life of the process, which for a process that creates and destroys workers would accumulate.
3. **If this is real, is the intended fix the obvious one?** Something like: check `ctx.Err()` at the top of the loop, and replace the trailing `time.Sleep(10 * time.Millisecond)` with a `select` on a timer and `ctx.Done()`. That would also make the retry interruptible, which I think is wanted regardless. Happy to send a PR with a test if the approach sounds right.

If I have misread any of this, please say so and I will drop it.

## Specifications

- Version: `main` at `8e2c89c7` (v1.46.0). The code is unchanged as far back as I looked.
- Platform: macOS (arm64), go1.25. Nothing platform-specific about it.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,13 @@ to docs, or any other relevant information.
   newly encountered `workflow.GetVersion` call. This supports gradual rollout of a new
   `GetVersion` call before activating its new behavior.
 
+### Fixed
+
+- Resource-based tuner: `ReserveSlot` now honors context cancellation while the resource controller is
+  declining slots. Previously the retry loop observed the context only while the ramp throttle was making
+  the caller wait, so a poller goroutine could outlive worker shutdown, keeping the worker's stop
+  `WaitGroup` from draining and continuing to sample system resources for the life of the process.
+
 ## [1.46.0] - 2026-07-07
 
 ### Fixed

--- a/internal/resource_tuner.go
+++ b/internal/resource_tuner.go
@@ -16,6 +16,10 @@ const (
 	resourceSlotsMemUsage = "temporal_resource_slots_mem_usage"
 )
 
+// retryReserveInterval is how long ReserveSlot waits before re-asking the controller for a slot
+// after it has declined one.
+const retryReserveInterval = 10 * time.Millisecond
+
 // SysInfoProvider implementations provide information about system resources.
 //
 // Exposed as: [go.temporal.io/sdk/worker.SysInfoProvider]
@@ -180,6 +184,9 @@ func NewResourceBasedSlotSupplier(
 
 func (r *ResourceBasedSlotSupplier) ReserveSlot(ctx context.Context, info SlotReservationInfo) (*SlotPermit, error) {
 	for {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
 		if info.NumIssuedSlots() < r.options.MinSlots {
 			return &SlotPermit{}, nil
 		}
@@ -201,7 +208,11 @@ func (r *ResourceBasedSlotSupplier) ReserveSlot(ctx context.Context, info SlotRe
 		if maybePermit != nil {
 			return maybePermit, nil
 		}
-		time.Sleep(10 * time.Millisecond)
+		select {
+		case <-time.After(retryReserveInterval):
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
 	}
 }
 

--- a/internal/resource_tuner_cancel_test.go
+++ b/internal/resource_tuner_cancel_test.go
@@ -1,0 +1,103 @@
+package internal
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/sdk/internal/common/metrics"
+	ilog "go.temporal.io/sdk/internal/log"
+	"go.temporal.io/sdk/log"
+)
+
+// reserveCallCounter counts NumIssuedSlots calls, so a test can observe that ReserveSlot has entered
+// its retry loop without resorting to a fixed sleep.
+type reserveCallCounter struct {
+	calls atomic.Int64
+}
+
+func (c *reserveCallCounter) TaskQueue() string { return "test-tq" }
+func (c *reserveCallCounter) TaskQueueKind() enumspb.TaskQueueKind {
+	return enumspb.TASK_QUEUE_KIND_NORMAL
+}
+func (c *reserveCallCounter) WorkerBuildId() string           { return "" }
+func (c *reserveCallCounter) WorkerIdentity() string          { return "" }
+func (c *reserveCallCounter) NumIssuedSlots() int             { c.calls.Add(1); return 10 }
+func (c *reserveCallCounter) Logger() log.Logger              { return ilog.NewNopLogger() }
+func (c *reserveCallCounter) MetricsHandler() metrics.Handler { return metrics.NopHandler }
+
+// TestReserveSlotRejectsAlreadyCancelledContext covers the MinSlots fast path, which returns a permit
+// without consulting the controller. semaphore.Acquire, which FixedSizeSlotSupplier reserves through,
+// fails a cancelled context even when it could acquire without blocking; this keeps the two suppliers
+// consistent.
+func TestReserveSlotRejectsAlreadyCancelledContext(t *testing.T) {
+	rcOpts := DefaultResourceControllerOptions()
+	rcOpts.InfoSupplier = &FakeSystemInfoSupplier{memUse: 0.1, cpuUse: 0.1}
+
+	supplier, err := NewResourceBasedSlotSupplier(NewResourceController(rcOpts),
+		ResourceBasedSlotSupplierOptions{MinSlots: 100, MaxSlots: 1000, RampThrottle: 0})
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	permit, err := supplier.ReserveSlot(ctx, &reserveCallCounter{}) // NumIssuedSlots is 10, under MinSlots
+	require.ErrorIs(t, err, context.Canceled)
+	require.Nil(t, permit)
+}
+
+// TestReserveSlotHonorsContextCancellation covers worker shutdown while the controller is declining
+// slots. Both cases below skip the ramp wait, which used to be the only place the context was
+// observed, leaving the retry loop with no cancellation point at all.
+func TestReserveSlotHonorsContextCancellation(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		ramp time.Duration
+	}{
+		// The default for workflow slot suppliers: the ramp block is skipped entirely.
+		{"no ramp throttle", 0},
+		// A ramp is configured, but the last issuance is old enough that no wait is required.
+		{"ramp throttle already elapsed", 50 * time.Millisecond},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			rcOpts := DefaultResourceControllerOptions()
+			rcOpts.MemTargetPercent = 0.8
+			rcOpts.CpuTargetPercent = 0.9
+			// Over the memory target, so the controller always declines and TryReserveSlot always
+			// returns nil. This is the sustained-pressure case the tuner exists to handle.
+			rcOpts.InfoSupplier = &FakeSystemInfoSupplier{memUse: 0.95, cpuUse: 0.95}
+
+			supplier, err := NewResourceBasedSlotSupplier(NewResourceController(rcOpts),
+				ResourceBasedSlotSupplierOptions{MinSlots: 0, MaxSlots: 1000, RampThrottle: tc.ramp})
+			require.NoError(t, err)
+
+			info := &reserveCallCounter{} // NumIssuedSlots is 10, past MinSlots
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			reserveErr := make(chan error, 1)
+			go func() {
+				_, err := supplier.ReserveSlot(ctx, info)
+				reserveErr <- err
+			}()
+
+			// Each retry iteration calls NumIssuedSlots once. Waiting for a second call means the
+			// loop is spinning, so cancelling now exercises the retry path rather than entry.
+			require.Eventually(t, func() bool {
+				return info.calls.Load() > 1
+			}, 2*time.Second, time.Millisecond, "ReserveSlot did not enter its retry loop")
+
+			cancel()
+
+			select {
+			case err := <-reserveErr:
+				require.ErrorIs(t, err, context.Canceled)
+			case <-time.After(2 * time.Second):
+				t.Fatal("ReserveSlot did not return after its context was cancelled")
+			}
+		})
+	}
+}

